rfcs/0005-nix-encryption.md

```diff
@@ -0,0 +1,273 @@
+---
+feature: nix-encryption
+start-date: 2016-03-28
+author: Eelco Dolstra
+co-authors:
+related-issues: https://github.com/NixOS/nix/issues/8, https://github.com/NixOS/nix/pull/329
+---
+
+# Summary
+[summary]: #summary
+
+We currently lack a way to store secret information in the Nix
+store. The proposal is to add add a builtin function to Nix to allow
+secrets to be encrypted at evaluation time with a key. At runtime,
+files containing encrypted data can be decrypted using the same key.
+
+# Motivation
+[motivation]: #motivation
+
+It is often necessary to store secrets such as passwords or SSL
+certificates in NixOS system closures. Unfortunately, the Nix store is
+world-readable, so this is currently not possible in a secure way;
+anybody who gains access to the filesystem of a NixOS machine can read
+any file in its Nix store. Nevertheless, lots of NixOS modules *do*
+store secrets in the Nix store (see
+https://github.com/NixOS/nixpkgs/issues/24288), many of them lacking
+warnings about the danger of doing so.
+
+# Detailed design
+[design]: #detailed-design
+
+The proposal consists of the following components:
+
+* A builtin function `encryptString :: Path -> String -> String` that
+  encrypts a string using a symmetric key stored in the specified
+  path. The encryption is done using libsodium's authenticated
+  encryption function `crypto_secretbox_easy()`. The resulting string
+  has the format `<{|nixcrypt:base64-encoded-data|}>`, where
+  `base64-encoded-data` is the base-64-encoded output of
+  `crypto_secretbox_easy()`, along with the nonce used.
+
+* A command `nix decrypt` to decrypt files containing encrypted
+  strings produced by `encryptString`. This command searches for
+  `<{|nixcrypt:...|}>` fragments and decrypts them using a decryption
+  key specified on the command line. Any other data is copied
+  verbatim.
+
+* A command `nix generate-key` to generate a key in the format
+  expected by `encryptString` and `nix decrypt`.
+
+A typical usage in a module NixOS would look like this:
+
+```
+let
+  configFile = pkgs.writeText "foo.conf"
+    ''
+      # Store the password of the foo service in encrypted form.
+      password=${builtins.encryptString <nixos-store-key> cfg.password}
+    '';
+in {
+
+  systemd.services.foo = {
+    preStart =
+      ''
+        # Decrypt the configuration file.
+        ${config.nix.package}/bin/nix-store \
+          --decrypt ${toString <nixos-store-key>} \
+          ${configFile} > /run/secret/foo.conf
```

Comment:
Exactly my thought.   Currently,  `nixops` supports delivering keys out of band a well and saves them in `/run/keys` .  I think the actual hard part is not solved by this RFC as it gives the same level of flexibility as what `nixops` currently already delivers!    The real hard part is that we need to rewrite dozens of  NixOS modules  that  directly  store secrets such that they can retrieve their secrets from disk.  And we should discourage or maybe totally deprecate any kind of other usage ...

Perhaps a larger part of the proposal isn't actually the encryption builtin,  but  actually rewriting modules in such way that they support this paradigm (and the nixops / vault paradigm)